src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,369 @@
+            TypeSymbol leftTupleType = MakeConvertedType(operators, node.Left, left, isNullable, compilation, diagnostics, isRight: false);
+            TypeSymbol rightTupleType = MakeConvertedType(operators, node.Right, right, isNullable, compilation, diagnostics, isRight: true);
+
+            return new TupleBinaryOperatorInfo.Nested(operators, leftTupleType, rightTupleType);
+        }
+
+        private static int? GetTupleCardinality(BoundExpression expr)
+        {
+            if (expr.Kind == BoundKind.TupleLiteral)
+            {
+                var tuple = (BoundTupleLiteral)expr;
+                return tuple.Arguments.Length;
+            }
+
+            TypeSymbol type = expr.Type;
+            if (type is null)
+            {
+                return null;
+            }
+
+            type = type.StrippedType();
+
+            if (type.IsTupleType)
+            {
+                return type.TupleElementTypes.Length;
+            }
+
+            return null;
+        }
+
+        private static ImmutableArray<BoundExpression> GetTupleArgumentsOrPlaceholders(BoundExpression expr)
+        {
+            if (expr.Kind == BoundKind.TupleLiteral)
+            {
+                return ((BoundTupleLiteral)expr).Arguments;
+            }
+
+            // placeholder bound nodes with the proper types are sufficient to bind the element-wise binary operators
+            return expr.Type.StrippedType().TupleElementTypes
+                .SelectAsArray((t, s) => (BoundExpression)new BoundTupleOperandPlaceholder(s, t), expr.Syntax);
+        }
+
+        /// <summary>
+        /// Make a tuple type (with appropriate nesting) from the types (on the left or on the right) collected
+        /// from binding element-wise binary operators.
+        /// </summary>
+        private TypeSymbol MakeConvertedType(ImmutableArray<TupleBinaryOperatorInfo> operators, CSharpSyntaxNode syntax,
+            ImmutableArray<BoundExpression> elements, bool isNullable, CSharpCompilation compilation, DiagnosticBag diagnostics, bool isRight)
+        {
+            ImmutableArray<TypeSymbol> convertedTypes = operators.SelectAsArray((o, r) => r ? o.RightConvertedType : o.LeftConvertedType, isRight);
+            ImmutableArray<Location> elementLocations = elements.SelectAsArray(e => e.Syntax.Location);
+
+            var tuple = TupleTypeSymbol.Create(locationOpt: null, elementTypes: convertedTypes,
+                elementLocations, elementNames: default, compilation,
+                shouldCheckConstraints: true, errorPositions: default, syntax, diagnostics);
+
+            if (!isNullable)
+            {
+                return tuple;
+            }
+
+            NamedTypeSymbol nullableT = GetSpecialType(SpecialType.System_Nullable_T, diagnostics, syntax);
+            return nullableT.Construct(tuple);
+        }
+    }
+
+    /// <summary>
+    /// A tree of binary operators
+    /// </summary>
+    internal abstract class TupleBinaryOperatorInfo
+    {
```

Comment:
Please move to its own source file.

Reply:
in BoundTree directory.

---
In reply to: [168053080](https://github.com/dotnet/roslyn/pull/23356#discussion_r168053080) [](ancestors = 168053080)